flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,211 @@
+#include "flecsi/data/topology.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/flog.hh"
+#include "flecsi/util/demangle.hh"
+
+#include <memory>
+#include <mutex>
+#include <string>
+
+namespace flecsi {
+namespace topo {
+struct global_base;
+}
+
+namespace exec {
+
+template<task_processor_type_t ProcessorType>
+struct task_prologue {
+
+protected:
+  // Those methods are "protected" because they are *only* called by
+  // flecsi::exec::prolog() which inherits from task_prologue.
+
+  // Patch up the "un-initialized" type conversion from future<R, index>
+  // to future<R, single> in the generic code.
+  template<typename R>
+  static void visit(future<R, exec::launch_type_t::single> & single,
+    future<R, exec::launch_type_t::index> & index) {
+    single = index.get(flecsi::run::context::instance().color());
+  }
+
+  template<typename T,
+    Privileges P,
+    class Topo,
+    typename Topo::index_space Space>
+  void visit(data::accessor<data::raw, T, P> & accessor,
+    const data::field_reference<T, data::raw, Topo, Space> & ref) {
+
+    // the spinlock-pool is used to protect the initialization of the
+    // futures that are associated with the storage for a field
+    using spinlock_pool =
+      ::hpx::util::spinlock_pool<::hpx::shared_future<void>>;
+
+    const field_id_t f = ref.fid();
+    auto & t = ref.topology();
+    data::region & reg = t.template get_region<Space>();
+    constexpr bool glob =
+      std::is_same_v<typename Topo::base, topo::global_base>;
+
+    auto & r = [&]() -> auto & {
+      if constexpr(glob) {
+        return *t;
+      }
+      else {
+        // The partition controls how much memory is allocated.
+        return t.template get_partition<Space>(f);
+      }
+    }
+    ();
+
+    // if the current argument as a valid future associated with it, then
+    // it has to be used a dependency for executing the current task
+    auto & field_future = r[f].future;
+    if(field_future.valid()) {
+      std::lock_guard l(spinlock_pool::spinlock_for(&field_future));
+      dependencies.push_back(field_future);
+    }
+
+    const auto storage = r.template get_storage<T, ProcessorType>(f);
+    if constexpr(glob) {
+      using data_type = ::hpx::serialization::serialize_buffer<T>;
+      auto comm = flecsi::run::context::instance().world_comm(
+        "prolog_visit" + std::to_string(f));
+      if(reg.ghost<privilege_pack<get_privilege(0, P), ro>>(f)) {
+        if(comm.is_root()) {
+          auto f = ::hpx::collectives::broadcast_to(comm,
+            data_type(storage.data(), storage.size(), data_type::reference));
+          f.get();
+        }
+        else {
+          auto f = ::hpx::collectives::broadcast_from<data_type>(comm);
+          auto && data = f.get();
+          assert(data.size() == storage.size());
+          std::move(data.begin(), data.begin() + data.size(), storage.data());
+        }
+      }
+    }
+    else {
+      reg.ghost_copy<P>(ref);
```

Comment:
Note that this can, even if `P` is all read-only, reentrantly post a task that _writes_ to the field (because it needs to update ghost values before the user task reads them).  That may not be a problem, but it has to be considered when constructing dependencies.